src/ppx_utils.ml

```diff
@@ -0,0 +1,51 @@
+open Import
+
+type kind =
+  | Structure
+  | Signature
+  | Unknown
+
+let relative_document_path ~document =
+  Workspace.asRelativePath ~pathOrUri:(`Uri (TextDocument.uri document))
+
+let project_root_path ~document =
+  let relative = relative_document_path ~document in
+  match Workspace.rootPath () with
+  | Some rootPath -> rootPath
+  | None -> raise (Failure ("Couldnt find root path for " ^ relative))
+
+let get_kind ~document =
+  let relative = relative_document_path ~document in
+  match String.sub ~pos:(String.length relative - 3) ~len:3 relative with
+  | ".ml" -> Structure
+  | "mli" -> Signature
+  | _ -> Unknown
+
+let get_pp_path ~(document : TextDocument.t) =
+  try
+    let relative = relative_document_path ~document in
+    let root = project_root_path ~document in
+    match get_kind ~document with
+    | Structure ->
+      root ^ "/_build/default/"
+      ^ String.sub ~pos:0 ~len:(String.length relative - 2) relative
+      ^ "pp.ml"
+    | Signature ->
+      root ^ "/_build/default/"
+      ^ String.sub ~pos:0 ~len:(String.length relative - 3) relative
+      ^ "pp.mli"
+    | Unknown -> failwith "Unknown file extension"
+  with
+  | Failure errorMsg -> errorMsg
```

Comment:
who is raising `Failure` here? This error should never really be caught.

Reply:
the `project_root_path` did in case where root wasn't found (when is that even possible?). Anyway, changed it to managing `string option`  returned by `Workspace.rootPath` in 3a57c06e207942b7bca2f7c88d0471cce3e70a5b